### PR TITLE
feat(compiler-cli): report error if undecorated class with Angular features is discovered

### PR DESCRIPTION
**Pull request by devversion**
* See individual commits.

### CODE REVIEW

**Comment by devversion** [comment]
Note: The components repo unit test job currently fails as a new compiler diagnostic has been introduced in this PR and errors would need to be solved in the components repo. 

Regardless of that failing job though, it would be good to receive feedback on the PR.

**Review by atscott — APPROVED** [review]
reviewed for: fw-router

File: packages/compiler-cli/src/ngtsc/annotations/src/diagnostics.ts [hunk 1]
```diff
@@ -78,6 +78,14 @@ export function getDirectiveDiagnostics(
   return diagnostics;
 }
 
+export function getUndecoratedClassWithAngularFeaturesDiagnostic(node: ClassDeclaration):
+    ts.Diagnostic {
+  return makeDiagnostic(
+      ErrorCode.UNDECORATED_CLASS_USING_ANGULAR_FEATURES, node.name,
+      `Class is using Angular features but is not decorated. Please add an explicit ` +
```
Comment (alxhub) [inline packages/compiler-cli/src/ngtsc/annotations/src/diagnostics.ts:85]: This is good for now, but we want to make this error message better before final.

1) the error should go on the feature in question, not the class identifier. So if you have an `@Input() field: ...` on the class, the error should go on the `@Input` decorator itself. Or the lifecycle method name, etc.

2) the error should include an extra pointer to the class name (a `relatedInfo` node) that suggests adding a decorator to the class itself. For `@Input` or any of the directive/component specific annotations, we should suggest adding `@Directive()`. For lifecycle hooks or DI annotations in the constructor, we should suggest adding any decorator.

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts [hunk 2]
```diff
@@ -905,6 +907,23 @@ runInEachFileSystem(os => {
               `static ɵdir: i0.ɵɵDirectiveDefWithMeta<TestBase, never, never, { "input": "input"; }, {}, never>;`);
     });
 
+    it('should error if undecorated class with Angular features has been discovered', () => {
```
Comment (alxhub) [inline packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:910]: Add a test for the lifecycle case as well.

**Review by alxhub — APPROVED** [review]
LGTM with a few comments, for the compiler piece only.

File: packages/compiler-cli/src/ngtsc/annotations/src/diagnostics.ts (hunk 1 above)
Comment (devversion) [reply packages/compiler-cli/src/ngtsc/annotations/src/diagnostics.ts:85]: (1) 👍 
(2) 👍 I did something like that initially but left that out as it felt out-of-scope. I can re-use that. https://gist.github.com/devversion/d412695321ca2a37fc59d84b6255f733 (patch is reversed)

Comment (devversion) [reply packages/compiler-cli/src/ngtsc/annotations/src/diagnostics.ts:85]: @alxhub Regarding (2), you mention DI annotations. This is not something that is handled yet. It sounds like we should though? I assume we can do that as a follow-up too.

**Review by AndrewKushnir — APPROVED** [review]
Reviewed for: fw-forms

File: packages/compiler-cli/src/ngtsc/annotations/src/diagnostics.ts (hunk 1 above)
Comment (kara) [reply packages/compiler-cli/src/ngtsc/annotations/src/diagnostics.ts:85]: +1 to what Alex said. I'm good with that being a follow-up.

**Review by kara — CHANGES_REQUESTED** [review]
Main things look good. Typos and tests

**Review by petebacondarwin — APPROVED** [review]
Reviewed for: fw-upgrade, docs-infra

File: packages/compiler-cli/test/ngtsc/ngtsc_spec.ts [hunk 3]
```diff
@@ -905,6 +907,44 @@ runInEachFileSystem(os => {
...
+
+        const errors = env.driveDiagnostics();
+        expect(errors.length).toBe(1);
+        expect(trim(errors[0].messageText as string))
+            .toContain(
+                'Class is using Angular features but is not decorated. Please add an explicit ' +
+                'Angular decorator.');
+      });
+
+      it('should error if directive lifecye hook has been discovered', () => {
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:928]: lol. Not sure how that one got cut off. Good catch.

**Review by kara — APPROVED** [review]
LGTM

**Review by IgorMinar — APPROVED** [review]
LGTM! Thank you!

Reviewed-for: global-approvers

**Pull request closed without merging**